### DISCUSSION TITLE
How to assert that a Relation Manager is visible on a specific page context

### DISCUSSION THREAD

**Discussion by emcarson** [filamentphp/filament#11883] [Help]
### Package

Panel builder

### Package Version

v3.2.50

### How can we help you?

Hello all,

As the title suggests, I need to verify if a Relation Manager is visible/hidden on the View page (I'm using policies to hide the RM for users with specific roles).

As per the documentation, I know I can test the view page with the following:

```php
$user = User::factory()->create();

$livewire = Livewire::test(ViewUser::class, ['record' => $user->getRouteKey()]);

$livewire->assertFormSet([
    'email' => $user->email,
]);
```

And I know I can test the Relation Manager like this:

```php
Livewire::test(TransactionsRelationManager::class, [
    'ownerRecord' => $user,
    'pageClass' => ViewUser::class,
])->assertSuccessful();
```

Despite providing the `ownerRecord` and `pageClass` for context, the `assertSuccessful()` always returns true. Like it's being tested in isolation instead of the page context.

Is there a way to check if the RM is really being displayed on the view page? (other than checking for a specific text on the view page class)

Thank you

**Comment by joseph-d** [comment]
> Is there a way to check if the RM is really being displayed on the view page? (other than checking for a specific text on the view page class)

Did you even find a way to check for specific text on the view page class though? I couldn't even figure out how to do that.

Based on my own experience, testing the relation manager with `assertSuccessful()` as you mentioned always passes. And if you move up a level to test the `EditUser` or `ViewUser` classes, something like `assertSee('Some text')` won't pick up any text on the relation manager because the relation manager is loaded afterwards as a different livewire component.

**Reply by emcarson** [reply]
Hello @joseph-d,

I can test that the RelationManager title is displayed by doing an `assertSeeText()` on the ViewPage, but it completely ignores context (it always returns true even though I'm hiding the RM for specific roles).

**Comment by joseph-d** [comment]
Hi @emcarson 

That's weird, my experience isn't quite the same as yours.

Like you said, I am getting universal passing when testing on the Relation Manager, for example like this (in Pest):

```php
        livewire(UserResource\RelationManagers\TransactionsRelationManager::class, [
            'ownerRecord' => $job,
            'pageClass' => ViewUser::class,
        ])
            ->assertSuccessful();
```

But if I test for the Relation Manager title on the View page it fails every time, even when the Relation Manager is there:

```php
    $dump = livewire(ViewUser::class, [
        'record' => $job->getRouteKey(),
    ])->assertSeeText("Transactions for this user"); // The actual title of the relation manager
```

This fails every time. And if I `dump($dump)` above it makes sense that it'd fail because the test really isn't there. In fact, even just viewing the browser source it makes sense that it'd fail because that text isn't present in the HTML document either. There's a 2nd request that goes to `http://localhost/livewire/update` which gets the relation manager and that title.

So the rabbit hole I've gone down is to try and figure out how to test the content of that `update` request. Or, even better, try and figure out if there's a way to test for that `update` being requested (maybe there's a specific event being fired we can test for?) because I've noticed on the pages where I've hidden the Relation Manager the request for `update` is never made so it feels like there's a solid thing we should be able to test for somehow.

But I'm still a bit puzzled about how your `assertSeeText()` can be passing for the relation manager title unless you're running the test on something slightly different than I am?

**Reply by emcarson** [reply]
That's strange @joseph-d 

I'm using PHPUnit, shouldn't make a difference I suppose.

I have the following heading defined in my RM:

```php
protected static ?string $title = 'My Wallet';
```

And the following assertion passes:

```php
Livewire::test(TransactionsRelationManager::class, [
            'ownerRecord' => $model,
            'pageClass' => $this->viewPage,
        ])->assertSuccessful()->assertSeeText('My Wallet');
```

If I `dd()` the Livewire component I get 4 matches of the text (a heading property and 3 HTML references, which I can also see when I view the source):

```json
        #heading: "My Wallet"
```

```html
         <!--[if BLOCK]><![endif]-->        <div class="grid gap-y-1">\n
                      <!--[if BLOCK]><![endif]-->                <h3\n
                              class="fi-ta-header-heading text-base font-semibold leading-6 text-gray-950 dark:text-white"\n
                          >\n
                              My Wallet\n
                          </h3>\n
                      <!--[if ENDBLOCK]><![endif]-->\n
          \n
                      <!--[if BLOCK]><![endif]--><!--[if ENDBLOCK]><![endif]-->\n
                  </div>\n
              <!--[if ENDBLOCK]><![endif]-->\n
```

BTW, I'm using Filament v3 and I'm also setting the following property of all my RMs:

```php
protected static bool $isLazy = false;
```

This forces the RM to load on page load, regardless it it's visible or not. I did this because I needed to refresh the RM programatically after a Page Action and that failed if the RM was out of view (by default it's loaded only when you scroll it into view the first time).

Maybe you can set your RMs like this, or maybe find a way to change `$isLazy` just for the tests.

Good luck!

**Comment by howdu** [comment]
The only way I could find to test the relation manager permissions is to check if it's visible when passing the `activeRelationManager` index and testing if relation manager component exists.

```php
it('can render options for a product', function () {
    $relationManager = ProductResource\RelationManagers\OptionsRelationManager::class;

    $product = Product::factory()
        ->create();

    $this->get(ProductResource::getUrl('edit', [
        'record' => $product->id,
        'activeRelationManager' => array_search(
            $relationManager,
            ProductResource::getRelations()
        ),
    ]))
        ->assertOk()
        ->assertSeeLivewire($relationManager);
```

**Reply by emcarson** [reply]
Nice one @howdu, thank you!

Edit: Unfortunately it ignores the fact that the relation manager is hidden for specific roles. I guess it's a good way to check when the RM is always visible.

**Comment by joseph-d** [comment]
@howdu - brilliant, thank you!

@emcarson - how are you showing/hiding the RM? I'm not using roles, I'm hiding based on the status on the record using this method on the RM class, but perhaps the same thing would work for you just with the status check changed to a role check?

```php
    public static function canViewForRecord(Model $ownerRecord, string $pageClass): bool
    {
        return $ownerRecord->status !== 'saved' && $ownerRecord->status !== 'new';
    }
```

**Reply by emcarson** [reply]
Thank you @joseph-d.

I'm hiding by implementing the `viewAny` policy for the RM's model. The system is becoming quite complex so I'm trying to centralize all of these checks in my policies.
